### DISCUSSION TITLE
[Question]: Implementation and Support for SP800_56A in OpenSSL

### DISCUSSION THREAD

**Discussion by alaakryeem** [openssl/openssl#22822] [Q&A]
I'm interested in using the [SP800_56A ](https://csrc.nist.gov/pubs/sp/800/56/a/r3/final) key-establishment scheme in OpenSSL, and I was wondering if anyone knows whether it's already implemented in OpenSSL and supported by the EVP APIs.

If it is supported, could someone please provide some guidance on how to invoke it? I've searched the OpenSSL documentation and forums, but haven't found a clear answer.

Thanks in advance for any help or insights you can provide!

**Comment by alaakryeem** [comment]
If anyone has any insights or suggestions, I would really appreciate it. 
Thank you in advance for your help.

**Comment by slontis** [comment]
We only supply the primitives required to support SP800-56A. The protocols are not defined in OpenSSL.
OpenSSL supports FFC DH, ECC CDH Keygen & Key Agreement. (For proper support the FIPS provider is recommended for keygen) 
The EVP_MAC API's support MACS such as HMAC, CMAC, KMAC
Proper Key Validation is done in FIPS provider using the EVP_PKEY_XXX_check() functions.

For SP800-56C the KDF's HKDF (2 step) & SSKDF (1 step) are supported.
